Add labelled switch Stack

Southporter opened the pull request:
This adds a second Stack VM that is a labeled switch/computed goto variant. For the purposes of this PR, I made it a second file to be able to compare the code and the differences. This is a pretty basic port. I left the logic as much the same as possible, so there could be some optimization that I am missing out on here.

Below are some perf numbers from running `poop` on the bench program.
<details>
 <summary>Release Fast</summary>
 <pre>
Benchmark 1 (1436 runs): zig-out/bin/bench_stack
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          3.41ms ±  838us    2.18ms … 11.2ms         42 ( 3%)        0%
  peak_rss           1.03MB ± 16.9KB     770KB … 1.04MB         23 ( 2%)        0%
  cpu_cycles         1.04M  ± 1.10M        0   … 3.62M           0 ( 0%)        0%
  instructions       3.96M  ± 4.22M        0   … 9.91M           0 ( 0%)        0%
  cache_references   1.10K  ± 1.04K        0   … 3.53K           0 ( 0%)        0%
  cache_misses       75.0   ±  152         0   … 1.66K          86 ( 6%)        0%
  branch_misses       970   ± 1.01K        0   … 2.92K           0 ( 0%)        0%
Benchmark 2 (1297 runs): zig-out/bin/bench_switch
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          3.79ms ±  792us    2.62ms … 10.6ms         25 ( 2%)        💩+ 10.9% ±  1.8%
  peak_rss           1.06MB ± 15.8KB     909KB … 1.06MB         21 ( 2%)        💩+  2.8% ±  0.1%
  cpu_cycles          890K  ± 1.07M        0   … 3.58M           0 ( 0%)        ⚡- 14.3% ±  7.9%
  instructions       3.23M  ± 3.92M        0   … 9.28M           0 ( 0%)        ⚡- 18.3% ±  7.7%
  cache_references   1.30K  ± 1.24K        0   … 3.33K           0 ( 0%)        💩+ 18.3% ±  7.8%
  cache_misses       93.4   ±  192         0   … 1.63K         108 ( 8%)        💩+ 24.6% ± 17.2%
  branch_misses       927   ± 1.05K        0   … 2.79K           0 ( 0%)          -  4.4% ±  7.9%
  </pre>
</details>
<details>
  <summary> Debug</summary>
  <pre>
Benchmark 1 (204 runs): zig-out/bin/bench_stack
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          24.5ms ± 4.41ms    18.8ms … 38.4ms          0 ( 0%)        0%
  peak_rss           3.77MB ± 96.9KB    3.56MB … 3.94MB          0 ( 0%)        0%
  cpu_cycles         28.8M  ± 3.58M     11.9M  … 33.1M          31 (15%)        0%
  instructions        104M  ± 11.8M     41.6M  …  109M          37 (18%)        0%
  cache_references    105K  ± 17.8K     60.4K  …  126K           0 ( 0%)        0%
  cache_misses       9.56K  ± 9.43K      539   … 42.9K          12 ( 6%)        0%
  branch_misses      13.0K  ± 1.88K     6.40K  … 16.1K           6 ( 3%)        0%
Benchmark 2 (191 runs): zig-out/bin/bench_switch
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          26.1ms ± 4.65ms    19.2ms … 39.7ms          1 ( 1%)        💩+  6.6% ±  3.6%
  peak_rss           3.81MB ± 51.7KB    3.69MB … 3.94MB         66 (35%)          +  1.2% ±  0.4%
  cpu_cycles         27.9M  ± 5.31M      407K  … 32.2M          32 (17%)          -  3.1% ±  3.1%
  instructions       95.5M  ± 17.8M      332K  …  103M          39 (20%)        ⚡-  8.4% ±  2.8%
  cache_references    102K  ± 20.5K     9.03K  …  125K           1 ( 1%)          -  2.9% ±  3.6%
  cache_misses       10.9K  ± 10.1K      502   … 52.1K          19 (10%)          + 13.8% ± 20.2%
  branch_misses      12.8K  ± 2.44K      688   … 16.2K          11 ( 6%)          -  1.8% ±  3.3%
  </pre>
</details>

Based on this, it looks like it runs a little slower (at least on x86_64 Linux)

Closes #58 if merged

Southporter commented:
I was not expecting this to be merged as is, more to show the work and figure out how to proceed. I think i can get it to share most of the code instead of all the duplication. The main challenge will be the Opcode logic, but we should be able to share that with some inline functions. 
My main question here was whether you want to have 2 options for stack vms so that people can pick between them or if you only want one stack vm that fits the most targets.
I'll start working on the 2 separate stack vms route and get them sharing as much as possible, unless you want to go in another direction.

rdunnington commented:
Ok sounds good. I think since computed goto is slightly slower it's probably worth keeping separate. I agree inline functions are probably the way to go. Fwiw when I eventually resume work on the register backend, it'll probably make it a bit easier to port the opts to that VM anyway.

Southporter commented:
@rdunnington I think this is ready for review now. Profiling benchmarks are about the same. I moved the actual operations out to a separate file to make the rewrite a little easier. 

I made the implementation of the interpreter loop compile time dependent. That makes it easier to run the tests as we don't have to pass in a value at runtime for all the tests. I think this also makes more sense since I don't anticipate many people will want to run 2 different kinds of interpreter loops in one application and thus will allow the compiler to trim out the alternative implementation to prevent bloating the binary.

I added compile time defaults for MIPS to use the labeled switch, so this also closes #57 at the same time.
